Common/Orders/VwapOrder.cs

```diff
@@ -0,0 +1,72 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using QuantConnect.Interfaces;
+using QuantConnect.Securities;
+
+namespace QuantConnect.Orders
+{
+    /// <summary>
+    /// Guaranteed Volume Weighted Average Price (VWAP) order type
+    /// </summary>
+    public class VwapOrder : Order
+    {
+        /// <summary>
+        /// Guaranteed VWAP Order Type
+        /// </summary>
+        public override OrderType Type => OrderType.Vwap;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VwapOrder"/> class.
+        /// </summary>
+        public VwapOrder()
+        {
+        }
+
+        /// <summary>
+        /// Intiializes a new instance of the <see cref="VwapOrder"/> class.
+        /// </summary>
+        /// <param name="symbol">The security's symbol being ordered</param>
+        /// <param name="quantity">The number of units to order</param>
+        /// <param name="time">The current time</param>
+        /// <param name="tag">A user defined tag for the order</param>
+        /// <param name="properties">The order properties for this order</param>
+        public VwapOrder(Symbol symbol, decimal quantity, DateTime time, string tag = "", IOrderProperties properties = null)
+            : base(symbol, quantity, time, tag, properties)
+        {
+        }
+
+        /// <summary>
+        /// Gets the order value in units of the security's quote currency
+        /// </summary>
+        /// <param name="security">The security matching this order's symbol</param>
+        protected override decimal GetValueImpl(Security security)
+        {
+            return Quantity * security.Price;
```

Comment:
Yea, I guess this is the best we can do -- this value ends up getting used for sufficient capital for order checks -- which brings up a great question -- how are we allocating margin for this order? For example, will the system let me submit a VWAP order (doesn't fill immediately so it won't drain my cash until it fills) -- and then I could submit a market order for my total portfolio value and if we're not checking for it, the infrastructure will happily produce both fills and I think we'll end up with negative margin once the VWAP order fills.

Reply:
I guess what I mean is that we don't 'section off' capital when you submit an order, instead we check if you have enough and then when it fills we take some capital from the portfolio. I'm not sure if we do a sufficient capital for order check based on the price in the fill. For example, `security.Price` at the close may be very low, but the VWAP may be higher, but if we calculate sufficient capital for order using `security.Price` then we'll get the wrong answer -- this may not easily be solved unless we added the VWAP to the security, so maybe for now we push this one.

Reply:
@jaredbroad -- do you have thoughts on this? the concern is (assuming 100K starting margin) if I submit a VWAP order for an estimated 100K at 9:00 AM we'll do a sufficient capital check (on what I assume to be yesterday's close), then let's say I purchase 100K of some other asset at 12:00 PM (noon), then at 4:00PM I have zero remaining margin, are we still going to process the fill which would lead the portfolio to be negative 100K margin. Or even without purchasing something else, consider I submit a VWAP order for 100% of my portfolio and the asset moves up 5%, will I just get fewer shares or would the order bounce? Is the quantity defined by the dollar value or by the actual quantity -- for example, are you purchasing 100K worth of some asset at whatever the VWAP may be (quantity is derived from dollar value and VWAP) or is the quantity a first class citizen, IOW, I'm purchasing 250 shares and it just so happens to be worth 100K? Also, should we move an estimated amount into pending cash or something similar so that the algorithm can't double spend?

Reply:
Interesting. Its an easy test - lets do a VWAP this morning and see.